Invitation.vue improvements

BacLuc opened the pull request:
This PR fixes the following points in #631
- [ ] Wenn man auf den Einladungslink in der E-Mail klickt, dann kommt man auf einer Seite mit folgenden Möglichkeiten:
    - [ ] Der InviteKey wurde gefunden, man ist nicht eingeloggt:
        - [x] Login -> funktioniert
        - [x] Reject , der Status wechselt auf LEFT
    - Der InviteKey wurde gefunden, man ist eingeloggt:
        - [x] displayName des momentan verwendeten Accounts
        - [x] Deny, der Status wechselt auf LEFT
- [x] ~~If logged in~~, display some kind of "Cancel" link on the invitation accept page that brings the user back to the Home page
- [x] Nicht erlauben, dass ein User zwei mal eine Invitation akzeptieren kann

BacLuc committed:
InvitationController: prohibit that a user can accept a second invitation

Issue: #631

BacLuc committed:
Invitation.vue: show displayName of authenticated user

Issue: #631

BacLuc committed:
Invitation.vue: if user is already part of the camp, show "Open Camp" instead of "Accept"

Issue: #631

BacLuc committed:
Invitation.vue: implement rejectInvitation

Also who a message if an error occurred.
Refactor multiple booleans to named states.

Issue: #631

BacLuc committed:
Invitation.vue: add back to login button

That the user can abort without accepting or rejecting.

Issue: #631

pmattmann commented:
Aufgezählte Punkte funktionieren alle.
Mir ist neu folgendes aufgefallen:

1) Unschöne Fehlermeldung:
Nicht verständlich für einfache Benutzer.

![image](https://user-images.githubusercontent.com/470237/106381482-4dc83c80-63b9-11eb-9315-8f811a16b808.png)



2) Wenn jemand eine Einladung abgelehnt hat (Status=left), kann dieser nicht erneut eingeladen werden.
Es erscheint dann die oben gezeigt Fehlermeldung

BacLuc commented:
> Aufgezählte Punkte funktionieren alle.
> Mir ist neu folgendes aufgefallen:
> 
>     1. Unschöne Fehlermeldung:
>        Nicht verständlich für einfache Benutzer.
> 
> 
> ![image](https://user-images.githubusercontent.com/470237/106381482-4dc83c80-63b9-11eb-9315-8f811a16b808.png)

Die Validation Message kommt momentan aus dem Backend. (keine Translation z.b.)
Mal schauen ob man das Zusammensetzen und übersetzen der Message einfach dem Frontend überlassen kann.

> 
>     1. Wenn jemand eine Einladung abgelehnt hat (Status=left), kann dieser nicht erneut eingeladen werden.
>        Es erscheint dann die oben gezeigt Fehlermeldung

Jup, das wird mit 
- [ ] Man sieht auf der Collaborators Seite auch die CampCollaborations mit Status Left (Dann weiss man, welche E-Mails man nicht mehr senden kann. Es hat einen Knopf daneben: E-Mail erneut senden.)
gelöst.

carlobeltrame commented on frontend/src/views/camp/Invitation.vue:
```diff
@@ -1,16 +1,29 @@
   <auth-container>
-    <div v-if="invitationFound !== false && campCollaborationOpen !== false">
+    <div v-if="[states.INIT,states.INVITATION_FOUND].includes(dialogState)">
+      <h1 v-if="isLoggedIn" class="display-1">{{ this.$tc('components.invitation.userWelcome') }} {{ userDisplayName }}</h1>
```
Das wäre ein perfektes Beispiel für eine Übersetzung mit Platzhalter.
```suggestion
      <h1 v-if="isLoggedIn" class="display-1">{{ this.$tc('components.invitation.userWelcome', { name: userDisplayName }) }}</h1>
```

BacLuc replied:
das ging vergessen, kommt später

BacLuc replied:
das klappt bei mir irgendwie nicht, können wir sonst mal noch schnell anschauen

carlobeltrame commented on frontend/src/views/camp/Invitation.vue:
```diff
@@ -54,31 +80,76 @@ export default {
-  data: () => ({ invitationFound: undefined }),
+  data: () => ({
+    states: {
+      INIT: {},
+      INVITATION_FOUND: {},
+      INVITATION_NOT_FOUND: {},
+      REJECTED: {},
+      ERROR: {}
```
Alle Daten in `data` werden von Vue reaktiv gemacht, also vorbereitet auf Mutationen und Observers. Konstanten gehören meiner Meinung nach stattdessen ausserhalb der Komponenten-Objektstruktur hin, auf Zeile 76.

Für diesen Fall hier habe ich in Vue aber eher öfter ein anderes Pattern gesehen als die Enum-artigen States die du verwendest. Wo immer möglich werden Probleme möglichst funktional statt imperativ gelöst:
* Status `INIT` wird errechnet als `this.campCollaboration()._meta.loading`, entweder inline oder in einer `computed`
* Status `INVITATION_FOUND` ist der Normalzustand, der angenommen wird wenn keine Spezialfälle greifen
* Für die Stati `INVITATION_NOT_FOUND` und `REJECTED` würde ich separate views mit je einer Route einrichten, und zum gegebenen Zeitpunkt dorthin redirecten (`this.$router.push(...)`)
* Status `ERROR` könnte man an einer `data`-Variablen `error` erkennen, die im Fehlerfall mit der Exception befüllt wird
* `dialogState` und `states` kann man schliesslich entfernen

Was denkst du zu dieser Idee?

BacLuc replied:
so war es vorher, war aber meiner meinung nach komplizierter, da die states dann teilweise kombiniert waren.
Wirklich funktional war es auch nicht, da es immer von state abhing, der direkt in der komponente mutable war.

Weiter gab es unschöne fehler in der konsole, wenn ich gewisse requests ausgeführt habe ohne error handler.
(this.api.get().profile(), this.campCollaboration()._meta.load)

Ich wollte die konstanten erst ausserhalb haben, aber dann kann man nicht im template darauf zugreifen.
Man könnte für jeden state eine Wrapper computed funktion haben, damit ist aber auch nicht wirklich geholfen.

Wollen wir das mal in einer session anschauen?

BacLuc replied:
Wechsel auf berechung im backend implementiert

BacLuc commented on frontend/src/views/camp/Invitation.vue:
```diff
@@ -91,7 +107,7 @@ export default {
     },
     profile () {
-      if (this.isLoggedIn !== true) {
+      if (this.states.INIT === this.dialogState) {
```
Nachdem das camp geladen wurde, wird hier trotzdem this.api.get().profile() aufgerufen, was zu einem Fehler in der Konsole führt, falls man nicht eingeloggt ist.

BacLuc replied:
Wechsel auf berechung im backend implementiert

BacLuc committed:
Invitation: rewrite that the backend calculates all needed data

Also move the rejected screen to a separate route.
Before the InvitationController returned an embedded
camp, which the user had no access to.
(Because he was not in the camp)
When the user then tried to fetch the campCollaborations of the camp,
the title of the camp, available before,
was not available anymore.
This was an inconsistent behaviour, so i refactored it.

Issue: #631

BacLuc commented:
Wir zügeln das ins backend

carlobeltrame commented on backend/module/eCampApi/src/eCampApi/V1/Rpc/Invitation/InvitationController.php:
```diff
@@ -120,13 +124,19 @@ public function accept($inviteKey): HalJsonModel {
         } catch (EntityNotFoundException $e) {
             throw new EntityNotFoundException('Not Found', 404);
+        } catch (EntityValidationException $e) {
+            $entityValidationException = new EntityValidationException('Failed to update CampCollaboration', 422);
```
Consider changing this message to something more understandable, like "Failed to accept invitation"

BacLuc replied:
Jup, in this case the messages are set also from the underlying layer.
But Invitation is better.

carlobeltrame commented on backend/module/eCampCore/src/Hydrator/InvitationHydrator.php:
```diff
@@ -0,0 +1,30 @@
+     * @param object $object
+     */
+    public function hydrate(array $data, $object): Invitation {
+        throw new \BadMethodCallException('this method is not implemented');
```
Consider changing this message to something more helpful for frontend developers, like "Invitations are read-only" or "Writing invitations is not implemented"

BacLuc replied:
done

carlobeltrame commented on backend/module/eCampCore/src/Repository/CampCollaborationRepository.php:
```diff
@@ -22,6 +24,24 @@ public function findByInviteKey(string $inviteKey): ?CampCollaboration {
+        $q = $this->createQueryBuilder('c');
+        $q->where('c.user = :user')
+            ->andWhere('c.camp = :camp')
+        ;
```
Does php-cs-fixer accept this? Consider putting the semicolon at the end of the previous line.

BacLuc replied:
php-cs-fixer wants it like that.
Normally in our code we don't chain the query parts and repeat the $q, but i find that unnecessary and like the fluent api.
(And i don't like in place mutations, which repeating $q uses)

carlobeltrame commented on backend/module/eCampCore/src/Service/InvitationService.php:
```diff
@@ -19,21 +20,46 @@ class InvitationService {
         $this->authenticationService = $authenticationService;
-        $this->campCollaborationRepository = $serviceUtils->emGetRepository(CampCollaboration::class);
+        /** @var CampCollaborationRepository $entityRepository */
+        $entityRepository = $serviceUtils->emGetRepository(CampCollaboration::class);
```
Maybe remove this unnecessary intermediate variable

BacLuc replied:
Then PHPStorm complains that i make an unsafe assignment (assigning an unknown return type of emGetRepository to a strictly typed field).
I rather add the type hint with a unnecessary intermediate variable.
PHPStorm does not complain about that. bö?

carlobeltrame commented on backend/module/eCampLib/test/V1/Rpc/Invitation/InvitationControllerTest.php:
```diff
@@ -53,27 +57,78 @@ public function testFindsInvitation() {
         {
            "self":{
-              "href":"http:///api/camp-collaborations/{$this->campCollaborationInvited->getId()}"
+              "href":"http:///api/invitation/find"
```
The API resource name should be in plural: `/invitations/find`. But we can fix that in a separate PR.

BacLuc replied:
Also for RCP Controllers?
-> separate PR

carlobeltrame commented on frontend/src/views/camp/Invitation.vue:
```diff
@@ -87,14 +119,24 @@ export default {
+        )
+    },
+    rejectInvitation () {
+      this.api.href(this.api.get()._meta.self, 'invitation', {
```
You can remove the `._meta.self` from this line and the parallel call in `acceptInvitation`. Most API methods in hal-json-vuex do that automatically when they get an object as first parameter.

BacLuc replied:
done, thank you

carlobeltrame commented on frontend/src/views/camp/Invitation.vue:
```diff
@@ -1,84 +1,116 @@
+  data: () => ({
+    error: false,
+    invitationFound: undefined,
+    rejected: false
```
I don't see any uses of this data property, can it be removed?

BacLuc replied:
yes, moved the error case also to a separate route.
The error case should not have worked too.

carlobeltrame commented on frontend/src/views/camp/Invitation.vue:
```diff
@@ -1,84 +1,116 @@
   <auth-container>
-    <div v-if="invitationFound !== false && campCollaborationOpen !== false">
-      <h1 class="display-1">{{ this.$tc('components.invitation.title') }} {{ camp ? camp.title : '' }}</h1>
+    <div v-if="ready">
```
You can put the `v-if` directly on the `auth-container` and remove the `div`. Then not even the empty white rectangle is shown before the invitation has loaded.

BacLuc replied:
done

carlobeltrame reviewed (changes requested):
I like it way better this way, thanks!

All my comments below are optional. But I found one case that doesn't work as intended: When the user is not logged in and visits the invitation link, then uses the "Login" button and logs in, he gets back to the not logged in view, because the invitation state from before logging in is still cached in the Vuex store. You can fix this by adding the following line as first line in the `mounted` hook of `Invitation.vue`:
```js
this.api.reload(this.invitation())
```
This is safe to do because hal-json-vuex will only ever send one concurrent network request for the same URI.

pmattmann commented on backend/module/eCampCore/src/Service/InvitationService.php:
```diff
@@ -19,21 +20,46 @@ class InvitationService {
+            $user = $this->userService->fetch($userId);
+            $userDisplayName = $user->getDisplayName();
+            $existingCampCollaboration = $this->campCollaborationRepository->findByUserAndCamp($user, $camp);
+            $userAlreadyInCamp = null != $existingCampCollaboration;
```
```suggestion
            $userAlreadyInCamp = null != $existingCampCollaboration && $existingCampCollaboration->isEstablished();
```

Beim zweiten Einladen (CampCollaboration.Status == Left) ist User bereits gesetzt.
Dann wird CampCollaboration gefunden, hat aber status "invited".
$userAlreadyInCamp meint wohl nur Established CampCollaborations.

![image](https://user-images.githubusercontent.com/470237/108624460-615d4500-7445-11eb-9f4e-51a691b151d6.png)

Obwohl CampCollaboration.Status == "invited"

pmattmann commented on backend/module/eCampCore/src/Service/InvitationService.php:
```diff
@@ -42,27 +68,41 @@ public function acceptInvitation(string $inviteKey, string $userId): CampCollabo
         }
+        $camp = $campCollaboration->getCamp();
+        $existingCampCollaboration = $this->campCollaborationRepository->findByUserAndCamp($user, $camp);
+        if (null != $existingCampCollaboration) {
```
```suggestion
        if (null != $existingCampCollaboration && $existingCampCollaboration->isEstablished()) {
```

Gleiches Problem hier;
Gemeint ist wohl $existingCampCollaboration mit Status "Established"

BacLuc committed:
Update backend/module/eCampCore/src/Service/InvitationService.php

Co-authored-by: Pirmin Mattmann <pimattmann@gmail.com>

BacLuc committed:
Update backend/module/eCampCore/src/Service/InvitationService.php

Co-authored-by: Pirmin Mattmann <pimattmann@gmail.com>
